jobs/kpi-forecasting/kpi_forecasting/models/funnel_forecast.py

```diff
@@ -1,370 +1,48 @@
-        Fit and save a Prophet model for each segment combination.
-
-        Args:
-            observed_df (pd.DataFrame): dataframe of observations.  Expected to have columns
-                specified in the segments section of the config,
-                submission_date column with unique dates corresponding to each observation and
-                y column containing values of observations
-        """
-        for segment_settings in self.segment_models:
-            parameters = self._auto_tuning(observed_df, segment_settings)
-
-            # Initialize model; build model dataframe
-            add_log_growth_cols = (
-                "growth" in parameters.keys() and parameters["growth"] == "logistic"
-            )
-            test_dat = self._build_train_dataframe(
-                observed_df, segment_settings, add_log_growth_cols
-            )
-            model = self._build_model(segment_settings, parameters)
-
-            model.fit(test_dat)
-            if add_log_growth_cols:
-                # all values in these colunns are the same
-                parameters["floor"] = test_dat["floor"].values[0]
-                parameters["cap"] = test_dat["cap"].values[0]
-
-            if "holidays" in parameters.keys():
-                parameters["holidays"] = (
-                    parameters["holidays"]["holiday"].unique().tolist()
-                )
-            segment_settings.trained_parameters = parameters
-            segment_settings.segment_model = model
-
-    def _get_crossvalidation_metric(
-        self, m: prophet.Prophet, cv_settings: dict
-    ) -> float:
+    def _get_crossvalidation_metric(self, m: ProphetForecast) -> float:
         """function for calculated the metric used for crossvalidation
 
         Args:
-            m (prophet.Prophet): Prophet model for crossvalidation
+            m (ProphetForecast): Prophet model for crossvalidation
             cv_settings (dict): settings set by segment in the config file
 
         Returns:
-            float: Metric where closer to zero means a better model
+            float: Metric which should always be positive and where smaller values
+                indicate better models
         """
-        df_cv = cross_validation(m, **cv_settings)
+        df_cv = cross_validation(m.model, **self.cv_settings)
 
         df_bias = df_cv.groupby("cutoff")[["yhat", "y"]].sum().reset_index()
         df_bias["pcnt_bias"] = df_bias["yhat"] / df_bias["y"] - 1
         # Prophet splits the historical data when doing cross validation using
         # cutoffs. The `.tail(3)` limits the periods we consider for the best
         # parameters to the 3 most recent cutoff periods.
-        return df_bias.tail(3)["pcnt_bias"].mean()
+        return np.abs(df_bias.tail(3)["pcnt_bias"].mean())
 
-    def _auto_tuning(
-        self, observed_df, segment_settings: SegmentModelSettings
-    ) -> Dict[str, float]:
+    def _auto_tuning(self, observed_df) -> ProphetForecast:
```

Comment:
This is a useful idea. My favorite model tuning fact is that random search is usually more efficient than grid search, especially across higher-dimensional spaces. The KPI Prophet models have been tuned with random search. If this method is computationally intensive or long-running, switching to random search could be a quick improvement.

- [A concise explanation on Stack Exchange](https://stats.stackexchange.com/a/209409)
- [Figure 11.2 in _Deep Learning_](https://www.deeplearningbook.org/contents/guidelines.html) 
<img width="476" alt="image" src="https://github.com/user-attachments/assets/9dbcec45-3d64-4127-878b-570f76d2105c">

Reply:
Yeah I think a big improvement would be to refactor to use a standardized parameter optimization library like Optuna or something so we'd get different methods for free